### PR TITLE
Add xenial and systemd support.

### PR DESCRIPTION
**Pull request by ronin13**
- Added unit file for zookeeper.service.
- Xenial related updates.
    + zookeeperd not required with systemd.

Tested with deployment branches and is essentially no-op on existing trusty nodes.

### CODE REVIEW

File: manifests/install.pp [hunk 1]
```diff
@@ -23,9 +23,21 @@
...
-    require => Package['zookeeper']
+
+  if $::lsbdistcodename == 'xenial' {
+      file { '/usr/lib/systemd/system/zookeeper.service':
+          mode    => '0664',
+          source  => 'puppet:///modules/zookeeper/zookeeper.service',
+          require => Package['zookeeper'],
+      } ~> exec { 'systemctl daemon-reload':
+            refreshonly => true,
+            path        => $::path,
```
Comment (solarkennedy) [inline manifests/install.pp]: This is strange. You shouldn't need to do that (path)

File: manifests/install.pp [hunk 2]
```diff
@@ -23,9 +23,21 @@
...
+      }
+  } else {
+    package { ['zookeeperd']: #init.d scripts for zookeeper
```
Comment (solarkennedy) [inline manifests/install.pp]: So to confirm, on xenail we 'dont' need this package at all? On xenail does *it* come with a systemd service file?

File: manifests/service.pp [hunk 3]
```diff
@@ -5,14 +5,28 @@
...
-  service { 'zookeeper':
-    ensure     => 'running',
-    hasstatus  => true,
-    hasrestart => true,
-    enable     => true,
-    require    => [
-      Package['zookeeperd'],
-      File["${cfg_dir}/zoo.cfg"]
-    ]
+  if $::lsbdistcodename == 'xenial' {
```
Comment (solarkennedy) [inline manifests/service.pp]: I don't think you need this at all, you should be able to get away with just "zookeeper" and let puppet autodetect the provider.

File: manifests/install.pp [hunk 4]
```diff
@@ -23,9 +23,21 @@
...
-  package { ['zookeeperd']: #init.d scripts for zookeeper
-    ensure  => $ensure,
-    require => Package['zookeeper']
+
+  if $::lsbdistcodename == 'xenial' {
+      file { '/usr/lib/systemd/system/zookeeper.service':
+          mode    => '0664',
+          source  => 'puppet:///modules/zookeeper/zookeeper.service',
+          require => Package['zookeeper'],
+      } ~> exec { 'systemctl daemon-reload':
```
[the remark this answers is not in the document]
Comment (ronin13) [reply manifests/install.pp]: And `systemctl daemon-reload` is idempotent in either case (exec or module) I believe.

[the remark this answers is not in the document]
Comment (solarkennedy) [reply manifests/install.pp]: Yes. But you should get it for more than just that. Just use `systemd::unit_file` and let the module handle it.

File: manifests/install.pp (hunk 2 above)
Comment (solarkennedy) [reply manifests/install.pp]: This surprises me. Are we trying to use trusty versions of these packages on xenial hosts?

File: manifests/service.pp (hunk 3 above)
Comment (solarkennedy) [reply manifests/service.pp]: You can do that, or make the `package[zookeeperd]` have a `before`.

File: manifests/install.pp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ronin13) [reply manifests/install.pp]: Ack, will try with that.

File: manifests/install.pp (hunk 2 above)
Comment (ronin13) [reply manifests/install.pp]: We are not using trusty version of package on xenial hosts. (We don't maintain the zookeeper package ourselves, we use whatever is the one for that distro, which is https://packages.ubuntu.com/xenial/zookeeper (3.4.8-1) for xenial).


And, neither zookeeperd nor any package in xenial relating to zookeeper, ships with a unit file for systemd:  https://packages.ubuntu.com/search?keywords=zookeeper&searchon=all&suite=xenial&section=all

File: manifests/install.pp (hunk 1 above)
Comment (ronin13) [reply manifests/install.pp]: Updated to use systemd::unit_file  where this is not required.

File: manifests/service.pp (hunk 3 above)
Comment (ronin13) [reply manifests/service.pp]: Updated/fixed this.  service.pp is not altered now except  the before clause in install.pp. and removal of package[zookeeperd] from service.pp.

File: files/zookeeper.service [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+[Service]
+User=zookeeper
+Group=zookeeper
+SyslogIdentifier=zookeeper
+ExecStart=/bin/bash -c 'source /etc/default/zookeeper && exec /usr/share/zookeeper/bin/zkServer.sh start-foreground'
+Restart=always
+# This is due to zookeeper exiting with 143 when SIGTERM is sent by init.
+SuccessExitStatus=143
+LimitNOFILE=16384
```
Comment (solarkennedy) [inline files/zookeeper.service]: Seems low

File: files/zookeeper.service [hunk 6]
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Zookeeper distributed coordination server
+After=network.target
+Wants=network.target
```
Comment (solarkennedy) [inline files/zookeeper.service]: Is this right? Both `After`/`Wants` = network.target?

Comment (ronin13) [reply files/zookeeper.service]: Let me double check this. I guess this means, it is dependent on network.target.

File: files/zookeeper.service (hunk 5 above)
Comment (ronin13) [reply files/zookeeper.service]: Ack, I was thinking of adding override (which can be added separate from the unit file) in the zookeeper_cluster module. I can update it here itself. Thanks.

File: files/zookeeper.service (hunk 6 above)
Comment (ronin13) [reply files/zookeeper.service]: As per https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Before= 

"It is a common pattern to include a unit name in both the After= and Requires= options, in which case the unit listed will be started before the unit that is configured with these options"

**Pull request merged**